passageflex/src/main/java/id/passage/passageflex/PasskeyUtils.kt

```diff
@@ -0,0 +1,154 @@
+        val getCredRequest = GetCredentialRequest(listOf(getCredOption))
+        // Show the user Credential Manager with option to login with a Passkey
+        return credentialManager.getCredential(activity, getCredRequest)
+    }
+
+    /**
+     * Get Create Credential Options JSON
+     *
+     * Extract the public key from Passage webauthnStart challenge and transform into a JSON string
+     * @param challenge The webauthn start challenge sent from Passage
+     * @return String
+     * @throws CredentialParsingException
+     */
+    internal fun getCreateCredentialOptionsJson(challenge: CredentialCreationChallenge?): String {
+        val publicKey =
+            challenge?.challenge?.publicKey
+                ?: throw CredentialParsingException(CHALLENGE_MISSING)
+        val moshi = Moshi.Builder().build()
+        val adapter = CredentialCreationPublicKeyJsonAdapter(moshi)
+        return adapter.toJson(publicKey)
+            ?: throw CredentialParsingException(CHALLENGE_PARSING_FAILED)
+    }
+
+    /**
+     * Get Create Credential Handshake Response
+     *
+     * Extract the CreateCredentialResponse from Credential Manager response data and transform
+     * into the Passage handshake response class used to finish webauthn registration.
+     * @param createCredentialResponse The response from Android's Credential Manager
+     * @return ProtocolCredentialCreationResponse
+     * @throws CredentialParsingException
+     */
+    internal fun getCreateCredentialHandshakeResponse(createCredentialResponse: CreateCredentialResponse): CredentialCreationResponse1 {
+        val handshakeResponseJson =
+            createCredentialResponse.data.getString(REGISTRATION_RESPONSE_BUNDLE_KEY).toString()
+        val moshi = Moshi.Builder().build()
+        val handshakeResponseAdapter = CredentialCreationResponse1JsonAdapter(moshi)
+        return handshakeResponseAdapter.fromJson(handshakeResponseJson)
+            ?: throw CredentialParsingException(CREDENTIAL_PARSING_FAILED)
+    }
+
+    /**
+     * Get Credential Options JSON
+     *
+     * Extract the public key from Passage webauthnStart challenge and convert into a JSON string
+     * @param challenge The webauthn start challenge sent from Passage
+     * @return String
+     * @throws CredentialParsingException
+     */
+    internal fun getCredentialOptionsJson(challenge: CredentialAssertionChallenge?): String {
+        val credOptions =
+            challenge?.challenge?.publicKey
+                ?: throw CredentialParsingException(CHALLENGE_MISSING)
+        val moshi = Moshi.Builder().build()
+        val credOptionsAdapter = ProtocolCredentialAssertionPublicKeyJsonAdapter(moshi)
+        // Passage API bug: Login API frequently returns challenge with non-url-safe characters
+        // "+" and "/" that cause "Bad Base 64" exception to be thrown by the credential manager.
+        val modifiedChallenge = credOptions.challenge.replace('+', '-').replace('/', '_')
```

Comment:
Hmm interesting. Is this something we should guard against in the JS sdk?

Reply:
You should be good in JS :)